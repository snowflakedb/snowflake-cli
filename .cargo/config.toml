# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Rust configuration for MAXIMUM x86_64 compatibility
# Use the most conservative settings possible to avoid "Illegal instruction" errors
# This is for CI x86_64 Linux builds, not cross-compilation

[build]
# Let cargo use the native target but with conservative flags

[target.x86_64-unknown-linux-gnu]
# Use only baseline x86_64 instruction set (SSE/SSE2 only)
# Disable all optional/newer instruction sets for maximum compatibility
rustflags = [
    "-C", "target-cpu=core2",
    "-C", "target-feature=-sse3,-ssse3,-sse4.1,-sse4.2,-popcnt,-avx,-avx2,-fma,-bmi1,-bmi2,-lzcnt,-movbe",
    "-C", "opt-level=2",
    "-C", "lto=false"
]
