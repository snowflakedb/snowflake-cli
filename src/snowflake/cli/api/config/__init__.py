# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Configuration package with support for legacy and next-generation config systems.

This package provides a feature flag mechanism to switch between the legacy
configuration system and the new config-ng system using the environment variable
SNOWFLAKE_CLI_CONFIG_NG.

Usage:
    # Enable new config system
    export SNOWFLAKE_CLI_CONFIG_NG=1

    # Use legacy config system (default)
    unset SNOWFLAKE_CLI_CONFIG_NG

Environment Variables:
    SNOWFLAKE_CLI_CONFIG_NG: Set to '1', 'true', 'yes', or 'on' to enable config-ng
"""

from __future__ import annotations

import os
from typing import Literal

# Environment variable to enable the new config-ng system
CONFIG_NG_ENV_VAR: Literal["SNOWFLAKE_CLI_CONFIG_NG"] = "SNOWFLAKE_CLI_CONFIG_NG"


def _is_config_ng_enabled() -> bool:
    """
    Check if the new config-ng system should be used.

    Returns:
        True if config-ng should be used, False for legacy system
    """
    return os.environ.get(CONFIG_NG_ENV_VAR, "").lower() in ("1", "true", "yes", "on")


# Conditionally import from legacy or config-ng based on environment variable
if _is_config_ng_enabled():
    # Import ONLY from the new config-ng system
    # This will fail if config-ng is not fully implemented, which is intentional
    # to help identify what needs to be implemented
    try:
        from .config_ng import *  # noqa: F403, F401
    except ImportError as e:
        # If config-ng is not fully implemented, fail fast with a clear error
        raise ImportError(
            f"Config-ng is enabled via {CONFIG_NG_ENV_VAR} but is not fully implemented. "
            f"Missing implementation: {e}. "
            f"Either implement the missing functionality in config_ng.py or "
            f"disable config-ng by unsetting {CONFIG_NG_ENV_VAR}."
        ) from e
else:
    # Use the legacy configuration system
    from .legacy import *  # noqa: F403, F401

# Export the feature flag utilities and factory method for external use
__all__ = ["_is_config_ng_enabled", "CONFIG_NG_ENV_VAR", "get_config_manager"]
