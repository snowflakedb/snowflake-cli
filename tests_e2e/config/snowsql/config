[connections]

accountname = "default_connection_account"
username = "default_connection_user"
host = "localhost"
databasename = "default_connection_database"
schemaname = "public"
warehousename = "default_connection_warehouse"
rolename = "accountadmin"


[connections.snowsql1]
accountname = "a1"
username = "u1"
host = "h1"
databasename = "d1"
schemaname = "public"
warehousename = "w1"
rolename = "r1"

[connections.snowsql2]
accountname = "a2"
username = "u2"
host = "h2"
databasename = "d2"
schemaname = "public"
warehousename = "w2"
rolename = "r2"
port = 1234

[connections.example]
accountname = accountname
username = username


[variables]
example_variable=27


[options]
auto_completion = True
log_file = /tmp/snowsql.log
log_level = DEBUG
timing = True
output_format = psql
key_bindings = emacs
repository_base_url = https://sfc-repo.snowflakecomputing.com/snowsql
