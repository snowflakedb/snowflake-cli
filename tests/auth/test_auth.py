from pathlib import Path
from tempfile import TemporaryDirectory
from textwrap import dedent
from unittest import mock

import pytest
from snowflake.cli.api.config import get_connection_dict
from snowflake.cli.api.secure_utils import file_permissions_are_strict

EXECUTE_QUERY = "snowflake.cli._plugins.auth.keypair.manager.AuthManager.execute_query"
OBJECT_EXECUTE_QUERY = (
    "snowflake.cli._plugins.object.manager.ObjectManager.execute_query"
)
CONNECTION = "snowflake.cli._plugins.auth.keypair.manager.AuthManager._conn"
CONNECT = "snowflake.connector.connect"

new_connection = "keypairconnection"
user_name = "test_user"


@mock.patch(EXECUTE_QUERY)
@mock.patch(OBJECT_EXECUTE_QUERY)
@mock.patch(CONNECT)
def test_setup(
    mock_connect,
    mock_object_execute_query,
    mock_execute_query,
    runner,
    mock_cursor,
    os_agnostic_snapshot,
):
    _mock_user_and_empty_public_keys(
        mock_connect, mock_object_execute_query, mock_cursor
    )

    with TemporaryDirectory() as tmp_dir:
        result = runner.invoke(
            ["auth", "keypair", "setup", "--format", "JSON"],
            input=f"Y\n{new_connection}\n4096\n{tmp_dir}\n\n",
        )

        tmp_path = Path(tmp_dir)
        private_key_path = tmp_path / f"{new_connection}.p8"
        public_key_path = tmp_path / f"{new_connection}.pub"
        assert result.exit_code == 0, result.output
        assert result.output == dedent(
            f"""\
            Create a new connection? [Y/n]: Y
            Enter connection name: {new_connection}
            Enter key length [2048]: 4096
            Enter output path [~/.ssh]: {tmp_path.absolute()}
            Enter private key passphrase []: 
            {{
                "message": "Setup completed."
            }}
            """
        )
        assert private_key_path.exists()
        assert file_permissions_are_strict(private_key_path)
        assert public_key_path.exists()
        assert file_permissions_are_strict(public_key_path)
        assert _call_contains(
            mock_execute_query, f"ALTER USER {user_name} SET RSA_PUBLIC_KEY="
        )


@mock.patch(EXECUTE_QUERY)
@mock.patch(OBJECT_EXECUTE_QUERY)
@mock.patch(CONNECT)
def test_setup_with_password(
    mock_connect,
    mock_object_execute_query,
    mock_execute_query,
    runner,
    mock_cursor,
    os_agnostic_snapshot,
):
    _mock_user_and_empty_public_keys(
        mock_connect, mock_object_execute_query, mock_cursor
    )

    with TemporaryDirectory() as tmp_dir:
        result = runner.invoke(
            ["auth", "keypair", "setup", "--output-path", tmp_dir],
            input=f"Y\n{new_connection}\n4096\n123\n",
        )

        tmp_path = Path(tmp_dir)
        private_key_path = tmp_path / f"{new_connection}.p8"
        public_key_path = tmp_path / f"{new_connection}.pub"
        assert result.exit_code == 0, result.output
        assert result.output == os_agnostic_snapshot
        assert private_key_path.exists()
        assert file_permissions_are_strict(private_key_path)
        assert "BEGIN ENCRYPTED PRIVATE KEY" in private_key_path.read_text()
        assert public_key_path.exists()
        assert file_permissions_are_strict(public_key_path)
        assert _call_contains(
            mock_execute_query, f"ALTER USER {user_name} SET RSA_PUBLIC_KEY="
        )


@mock.patch(EXECUTE_QUERY)
@mock.patch(OBJECT_EXECUTE_QUERY)
@mock.patch(CONNECT)
def test_setup_no_prompts(
    mock_connect,
    mock_object_execute_query,
    mock_execute_query,
    runner,
    mock_cursor,
    os_agnostic_snapshot,
):
    _mock_user_and_empty_public_keys(
        mock_connect, mock_object_execute_query, mock_cursor
    )

    with TemporaryDirectory() as tmp_dir:
        result = runner.invoke(
            [
                "auth",
                "keypair",
                "setup",
                "--key-length",
                "4096",
                "--output-path",
                tmp_dir,
                "--private-key-passphrase",
                "123",
            ],
            input=f"\n{new_connection}\n",
        )

        assert result.exit_code == 0, result.output
        assert result.output == os_agnostic_snapshot


@mock.patch(EXECUTE_QUERY)
@mock.patch(OBJECT_EXECUTE_QUERY)
@mock.patch(CONNECT)
def test_setup_connection_already_exists(
    mock_connect,
    mock_object_execute_query,
    mock_execute_query,
    runner,
    mock_cursor,
    os_agnostic_snapshot,
):
    _mock_user_and_empty_public_keys(
        mock_connect, mock_object_execute_query, mock_cursor
    )

    with TemporaryDirectory() as tmp_dir:
        result = runner.invoke(
            ["auth", "keypair", "setup", "--output-path", tmp_dir], input="\ndefault\n"
        )

        assert result.exit_code == 1, result.output
        assert result.output == os_agnostic_snapshot


@mock.patch(EXECUTE_QUERY)
@mock.patch(OBJECT_EXECUTE_QUERY)
@mock.patch(CONNECT)
def test_setup_create_output_directory_with_proper_privileges(
    mock_connect,
    mock_object_execute_query,
    mock_execute_query,
    runner,
    mock_cursor,
    os_agnostic_snapshot,
):
    _mock_user_and_empty_public_keys(
        mock_connect, mock_object_execute_query, mock_cursor
    )

    with TemporaryDirectory() as tmp_dir:
        tmp_path = Path(tmp_dir) / "some" / "subdirectory" / "location"
        result = runner.invoke(
            ["auth", "keypair", "setup", "--output-path", tmp_path],
            input=f"Y\n{new_connection}\n4096\n\n",
        )

        private_key_path = tmp_path / f"{new_connection}.p8"
        public_key_path = tmp_path / f"{new_connection}.pub"
        assert result.exit_code == 0, result.output
        assert result.output == os_agnostic_snapshot
        assert file_permissions_are_strict(tmp_path)
        assert private_key_path.exists()
        assert file_permissions_are_strict(private_key_path)
        assert public_key_path.exists()
        assert file_permissions_are_strict(public_key_path.resolve())
        assert _call_contains(
            mock_execute_query, f"ALTER USER {user_name} SET RSA_PUBLIC_KEY="
        )


@pytest.mark.parametrize(
    "key_value, key_2_value",
    [
        ("KEY", None),
        (None, "KEY"),
        ("KEY", "KEY"),
    ],
)
@mock.patch(EXECUTE_QUERY)
@mock.patch(OBJECT_EXECUTE_QUERY)
@mock.patch(CONNECT)
def test_setup_error_if_any_public_key_is_set(
    mock_connect,
    mock_object_execute_query,
    mock_execute_query,
    runner,
    mock_cursor,
    os_agnostic_snapshot,
    key_value,
    key_2_value,
):
    mock_connect.return_value.user = user_name
    mock_object_execute_query.return_value = mock_cursor(
        rows=[
            {"property": "RSA_PUBLIC_KEY", "value": key_value},
            {"property": "RSA_PUBLIC_KEY_2", "value": key_2_value},
        ],
        columns=[],
    )

    with TemporaryDirectory() as tmp_dir:
        tmp_path = Path(tmp_dir) / "some" / "subdirectory" / "location"
        result = runner.invoke(
            ["auth", "keypair", "setup", "--output-path", tmp_path],
            input=f"Y\n{new_connection}\n4096\n\n",
        )

        assert result.exit_code == 1, result.output
        assert result.output == os_agnostic_snapshot


@mock.patch(EXECUTE_QUERY)
@mock.patch(OBJECT_EXECUTE_QUERY)
@mock.patch(CONNECT)
def test_setup_overwrite_connection(
    mock_connect,
    mock_object_execute_query,
    mock_execute_query,
    runner,
    mock_cursor,
    test_snowcli_config,
    os_agnostic_snapshot,
):
    _mock_user_and_empty_public_keys(
        mock_connect, mock_object_execute_query, mock_cursor
    )

    with TemporaryDirectory() as tmp_dir:
        result = runner.invoke(
            ["auth", "keypair", "setup", "--output-path", tmp_dir], input="n\n\n\n"
        )

        tmp_path = Path(tmp_dir)
        private_key_path = tmp_path / "default.p8"
        assert result.exit_code == 0, result.output
        assert result.output == os_agnostic_snapshot
        connection_config = get_connection_dict("default")
        assert connection_config["authenticator"] == "SNOWFLAKE_JWT"
        assert connection_config["private_key_file"] == str(private_key_path.resolve())
        assert "password" not in connection_config.keys()


@mock.patch(EXECUTE_QUERY)
@mock.patch(OBJECT_EXECUTE_QUERY)
@mock.patch(CONNECT)
def test_rotate(
    mock_connect,
    mock_object_execute_query,
    mock_execute_query,
    runner,
    mock_cursor,
    os_agnostic_snapshot,
):
    _mock_user_and_public_key(mock_connect, mock_object_execute_query, mock_cursor)

    with TemporaryDirectory() as tmp_dir:
        runner.invoke(
            ["auth", "keypair", "setup"],
            input=f"Y\n{new_connection}\n4096\n{tmp_dir}\n\n",
        )

        result = runner.invoke(
            ["auth", "keypair", "rotate", "--format", "JSON"],
            input=f"Y\n{new_connection}\n4096\n{tmp_dir}\n\n",
        )

        tmp_path = Path(tmp_dir)
        private_key_path = tmp_path / f"{new_connection}.p8"
        public_key_path = tmp_path / f"{new_connection}.pub"
        assert result.exit_code == 0, result.output
        assert result.output == dedent(
            f"""\
            Create a new connection? [Y/n]: Y
            Enter connection name: {new_connection}
            Enter key length [2048]: 4096
            Enter output path [~/.ssh]: {tmp_path.absolute()}
            Enter private key passphrase []: 
            {{
                "message": "Rotate completed."
            }}
            """
        )
        assert private_key_path.exists()
        assert file_permissions_are_strict(private_key_path)
        assert public_key_path.exists()
        assert file_permissions_are_strict(public_key_path)
        assert _call_contains(
            mock_execute_query, f"ALTER USER {user_name} SET RSA_PUBLIC_KEY_2="
        )


@mock.patch(EXECUTE_QUERY)
@mock.patch(OBJECT_EXECUTE_QUERY)
@mock.patch(CONNECT)
def test_rotate_with_password(
    mock_connect,
    mock_object_execute_query,
    mock_execute_query,
    runner,
    mock_cursor,
    os_agnostic_snapshot,
):
    _mock_user_and_public_key(mock_connect, mock_object_execute_query, mock_cursor)

    with TemporaryDirectory() as tmp_dir:
        runner.invoke(
            ["auth", "keypair", "setup"],
            input=f"Y\n{new_connection}\n4096\n{tmp_dir}\n\n",
        )

        result = runner.invoke(
            ["auth", "keypair", "rotate", "--output-path", tmp_dir],
            input=f"Y\n{new_connection}\n4096\n123\n",
        )

        tmp_path = Path(tmp_dir)
        private_key_path = tmp_path / f"{new_connection}.p8"
        public_key_path = tmp_path / f"{new_connection}.pub"
        assert result.exit_code == 0, result.output
        assert result.output == os_agnostic_snapshot
        assert private_key_path.exists()
        assert file_permissions_are_strict(private_key_path)
        assert "BEGIN ENCRYPTED PRIVATE KEY" in private_key_path.read_text()
        assert public_key_path.exists()
        assert file_permissions_are_strict(public_key_path)
        assert _call_contains(
            mock_execute_query, f"ALTER USER {user_name} SET RSA_PUBLIC_KEY_2="
        )


@mock.patch(EXECUTE_QUERY)
@mock.patch(OBJECT_EXECUTE_QUERY)
@mock.patch(CONNECT)
def test_rotate_no_prompts(
    mock_connect,
    mock_object_execute_query,
    mock_execute_query,
    runner,
    mock_cursor,
    os_agnostic_snapshot,
):
    _mock_user_and_public_key(mock_connect, mock_object_execute_query, mock_cursor)

    with TemporaryDirectory() as tmp_dir:
        result = runner.invoke(
            [
                "auth",
                "keypair",
                "rotate",
                "--key-length",
                "4096",
                "--output-path",
                tmp_dir,
                "--private-key-passphrase",
                "123",
            ],
            input=f"\n{new_connection}\n",
        )

        assert result.exit_code == 0, result.output
        assert result.output == os_agnostic_snapshot


@mock.patch(EXECUTE_QUERY)
@mock.patch(OBJECT_EXECUTE_QUERY)
@mock.patch(CONNECT)
def test_rotate_connection_already_exists(
    mock_connect,
    mock_object_execute_query,
    mock_execute_query,
    runner,
    mock_cursor,
    os_agnostic_snapshot,
):
    _mock_user_and_public_key(mock_connect, mock_object_execute_query, mock_cursor)

    with TemporaryDirectory() as tmp_dir:
        result = runner.invoke(
            ["auth", "keypair", "rotate", "--output-path", tmp_dir], input="\ndefault\n"
        )

        assert result.exit_code == 1, result.output
        assert result.output == os_agnostic_snapshot


@mock.patch(EXECUTE_QUERY)
@mock.patch(OBJECT_EXECUTE_QUERY)
@mock.patch(CONNECT)
def test_rotate_no_public_key_set(
    mock_connect,
    mock_object_execute_query,
    mock_execute_query,
    runner,
    mock_cursor,
    os_agnostic_snapshot,
):
    _mock_user_and_empty_public_keys(
        mock_connect, mock_object_execute_query, mock_cursor
    )

    with TemporaryDirectory() as tmp_dir:
        runner.invoke(
            ["auth", "keypair", "setup"],
            input=f"Y\n{new_connection}\n4096\n{tmp_dir}\n\n",
        )

        result = runner.invoke(
            ["auth", "keypair", "rotate", "--output-path", tmp_dir],
            input=f"\n{new_connection}\n",
        )

        assert result.exit_code == 1, result.output
        assert result.output == os_agnostic_snapshot


@mock.patch(EXECUTE_QUERY)
@mock.patch(OBJECT_EXECUTE_QUERY)
@mock.patch(CONNECT)
def test_rotate_only_public_key_set(
    mock_connect,
    mock_object_execute_query,
    mock_execute_query,
    runner,
    mock_cursor,
    os_agnostic_snapshot,
):
    mock_connect.return_value.user = user_name
    mock_object_execute_query.return_value = mock_cursor(
        rows=[
            {"property": "RSA_PUBLIC_KEY", "value": "KEY"},
            {"property": "RSA_PUBLIC_KEY_2", "value": None},
        ],
        columns=[],
    )

    with TemporaryDirectory() as tmp_dir:
        runner.invoke(
            ["auth", "keypair", "setup"],
            input=f"Y\n{new_connection}\n4096\n{tmp_dir}\n\n",
        )

        result = runner.invoke(
            ["auth", "keypair", "rotate", "--output-path", tmp_dir],
            input=f"Y\n{new_connection}\n4096\n\n",
        )

        tmp_path = Path(tmp_dir)
        private_key_path = tmp_path / f"{new_connection}.p8"
        public_key_path = tmp_path / f"{new_connection}.pub"
        assert result.exit_code == 0, result.output
        assert result.output == os_agnostic_snapshot
        assert private_key_path.exists()
        assert file_permissions_are_strict(private_key_path)
        assert public_key_path.exists()
        assert file_permissions_are_strict(public_key_path)
        assert _call_contains(
            mock_execute_query, f"ALTER USER {user_name} SET RSA_PUBLIC_KEY_2="
        )


@pytest.mark.parametrize(
    "key_value, key_2_value",
    [
        (None, "KEY"),
        ("KEY", "KEY"),
    ],
)
@mock.patch(EXECUTE_QUERY)
@mock.patch(OBJECT_EXECUTE_QUERY)
@mock.patch(CONNECT)
def test_rotate_other_public_key_set_options(
    mock_connect,
    mock_object_execute_query,
    mock_execute_query,
    runner,
    mock_cursor,
    os_agnostic_snapshot,
    key_value,
    key_2_value,
):
    mock_connect.return_value.user = user_name
    mock_object_execute_query.return_value = mock_cursor(
        rows=[
            {"property": "RSA_PUBLIC_KEY", "value": key_value},
            {"property": "RSA_PUBLIC_KEY_2", "value": key_2_value},
        ],
        columns=[],
    )

    with TemporaryDirectory() as tmp_dir:
        runner.invoke(
            ["auth", "keypair", "setup"],
            input=f"Y\n{new_connection}\n4096\n{tmp_dir}\n\n",
        )

        result = runner.invoke(
            ["auth", "keypair", "rotate", "--output-path", tmp_dir],
            input=f"Y\n{new_connection}\n4096\n\n",
        )

        tmp_path = Path(tmp_dir)
        private_key_path = tmp_path / f"{new_connection}.p8"
        public_key_path = tmp_path / f"{new_connection}.pub"
        assert result.exit_code == 0, result.output
        assert result.output == os_agnostic_snapshot
        assert private_key_path.exists()
        assert file_permissions_are_strict(private_key_path)
        assert public_key_path.exists()
        assert file_permissions_are_strict(public_key_path)
        assert _call_contains(
            mock_execute_query, f"ALTER USER {user_name} SET RSA_PUBLIC_KEY="
        )
        assert _call_contains(
            mock_execute_query, f"ALTER USER {user_name} SET RSA_PUBLIC_KEY_2="
        )


@mock.patch(EXECUTE_QUERY)
@mock.patch(OBJECT_EXECUTE_QUERY)
@mock.patch(CONNECT)
def test_rotate_overwrite_connection(
    mock_connect,
    mock_object_execute_query,
    mock_execute_query,
    runner,
    mock_cursor,
    test_snowcli_config,
    os_agnostic_snapshot,
):
    _mock_user_and_public_key(mock_connect, mock_object_execute_query, mock_cursor)

    with TemporaryDirectory() as tmp_dir:
        runner.invoke(
            ["auth", "keypair", "setup"],
            input=f"Y\n{new_connection}\n4096\n{tmp_dir}\n\n",
        )

        result = runner.invoke(
            ["auth", "keypair", "rotate", "--output-path", tmp_dir], input="n\n\n\n"
        )

        tmp_path = Path(tmp_dir)
        private_key_path = tmp_path / "default.p8"
        assert result.exit_code == 0, result.output
        assert result.output == os_agnostic_snapshot
        connection_config = get_connection_dict("default")
        assert connection_config["authenticator"] == "SNOWFLAKE_JWT"
        assert connection_config["private_key_file"] == str(private_key_path.resolve())
        assert "password" not in connection_config.keys()


@mock.patch(OBJECT_EXECUTE_QUERY)
@mock.patch(CONNECT)
def test_list(
    mock_connect, mock_execute_query, runner, mock_cursor, os_agnostic_snapshot
):
    mock_connect.return_value.user = user_name
    mock_execute_query.return_value = mock_cursor(
        rows=[
            {
                "property": "NAME",
                "value": "test_user",
                "default": None,
                "description": "Name",
            },
            {
                "property": "RSA_PUBLIC_KEY",
                "value": "-----BEGIN PUBLIC KEY-----",
                "default": None,
                "description": "RSA public key of the user",
            },
            {
                "property": "RSA_PUBLIC_KEY_FP",
                "value": "SHA256",
                "default": None,
                "description": "Fingerprint of user's RSA public key.",
            },
            {
                "property": "RSA_PUBLIC_KEY_LAST_SET_TIME",
                "value": "2025-02-17 12:53:51.212",
                "default": None,
                "description": "The timestamp",
            },
            {
                "property": "RSA_PUBLIC_KEY_2",
                "value": None,
                "default": None,
                "description": "Second RSA public key of the user",
            },
            {
                "property": "RSA_PUBLIC_KEY_2_FP",
                "value": None,
                "default": None,
                "description": "Fingerprint of user's second RSA public key.",
            },
            {
                "property": "RSA_PUBLIC_KEY_2_LAST_SET_TIME",
                "value": None,
                "default": None,
                "description": "The timestamp",
            },
            {
                "property": "COMMENT",
                "value": None,
                "default": None,
                "description": "user comment",
            },
        ],
        columns=[],
    )

    result = runner.invoke(["auth", "keypair", "list"])

    assert result.exit_code == 0, result.output
    assert result.output == os_agnostic_snapshot


@pytest.mark.parametrize("key", ["RSA_PUBLIC_KEY", "RSA_PUBLIC_KEY_2"])
@mock.patch(EXECUTE_QUERY)
@mock.patch(CONNECT)
def test_remove(
    mock_connect, mock_execute_query, runner, mock_cursor, os_agnostic_snapshot, key
):
    mock_connect.return_value.user = user_name
    mock_execute_query.return_value = mock_cursor(
        rows=[["Statement executed successfully."]], columns=["status"]
    )
    result = runner.invoke(["auth", "keypair", "remove", "--key-id", key])

    assert result.exit_code == 0, result.output
    assert result.output == os_agnostic_snapshot
    mock_execute_query.assert_called_once_with(f"ALTER USER test_user UNSET {key}")


def _mock_user_and_empty_public_keys(
    mock_connect, mock_object_execute_query, mock_cursor
):
    mock_connect.return_value.user = user_name
    mock_object_execute_query.return_value = mock_cursor(
        rows=[
            {"property": "RSA_PUBLIC_KEY", "value": None},
            {"property": "RSA_PUBLIC_KEY_2", "value": None},
        ],
        columns=[],
    )


def _mock_user_and_public_key(mock_connect, mock_object_execute_query, mock_cursor):
    mock_connect.return_value.user = user_name
    mock_object_execute_query.return_value = mock_cursor(
        rows=[
            {"property": "RSA_PUBLIC_KEY", "value": "KEY"},
            {"property": "RSA_PUBLIC_KEY_2", "value": None},
        ],
        columns=[],
    )


def _call_contains(mock, search_string: str) -> bool:
    for call in mock.mock_calls:
        if search_string in str(call):
            return True
    return False
