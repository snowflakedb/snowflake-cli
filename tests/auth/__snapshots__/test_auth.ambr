# serializer version: 1
# name: test_list
  '''
  +------------------------------------------------------------------------------+
  | property             | value                | default | description          |
  |----------------------+----------------------+---------+----------------------|
  | RSA_PUBLIC_KEY       | -----BEGIN PUBLIC    | None    | RSA public key of    |
  |                      | KEY-----             |         | the user             |
  | RSA_PUBLIC_KEY_FP    | SHA256               | None    | Fingerprint of       |
  |                      |                      |         | user's RSA public    |
  |                      |                      |         | key.                 |
  | RSA_PUBLIC_KEY_LAST_ | 2025-02-17           | None    | The timestamp        |
  | SET_TIME             | 12:53:51.212         |         |                      |
  | RSA_PUBLIC_KEY_2     | None                 | None    | Second RSA public    |
  |                      |                      |         | key of the user      |
  | RSA_PUBLIC_KEY_2_FP  | None                 | None    | Fingerprint of       |
  |                      |                      |         | user's second RSA    |
  |                      |                      |         | public key.          |
  | RSA_PUBLIC_KEY_2_LAS | None                 | None    | The timestamp        |
  | T_SET_TIME           |                      |         |                      |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_remove[RSA_PUBLIC_KEY]
  '''
  +-------------------------------------------+
  | key    | value                            |
  |--------+----------------------------------|
  | status | Statement executed successfully. |
  +-------------------------------------------+
  
  '''
# ---
# name: test_remove[RSA_PUBLIC_KEY_2]
  '''
  +-------------------------------------------+
  | key    | value                            |
  |--------+----------------------------------|
  | status | Statement executed successfully. |
  +-------------------------------------------+
  
  '''
# ---
# name: test_rotate_connection_already_exists
  '''
  Create a new connection? [Y/n]: 
  Enter connection name: default
  Enter key length [2048]: 
  Enter private key passphrase []: 
  +- Error ----------------------------------------------------------------------+
  | Connection with name default already exists.                                 |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_rotate_no_prompts
  '''
  Create a new connection? [Y/n]: 
  Enter connection name: keypairconnection
  Set the `PRIVATE_KEY_PASSPHRASE` environment variable before using the connection.
  Rotate completed.
  
  '''
# ---
# name: test_rotate_no_public_key_set
  '''
  Create a new connection? [Y/n]: 
  Enter connection name: keypairconnection
  Enter key length [2048]: 
  Enter private key passphrase []: 
  +- Error ----------------------------------------------------------------------+
  | Connection with name keypairconnection already exists.                       |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_rotate_only_public_key_set
  '''
  Create a new connection? [Y/n]: Y
  Enter connection name: keypairconnection
  Enter key length [2048]: 4096
  Enter private key passphrase []: 
  Rotate completed.
  
  '''
# ---
# name: test_rotate_other_public_key_set_options[KEY-KEY]
  '''
  Create a new connection? [Y/n]: Y
  Enter connection name: keypairconnection
  Enter key length [2048]: 4096
  Enter private key passphrase []: 
  Rotate completed.
  
  '''
# ---
# name: test_rotate_other_public_key_set_options[None-KEY]
  '''
  Create a new connection? [Y/n]: Y
  Enter connection name: keypairconnection
  Enter key length [2048]: 4096
  Enter private key passphrase []: 
  Rotate completed.
  
  '''
# ---
# name: test_rotate_overwrite_connection
  '''
  Create a new connection? [Y/n]: n
  Enter key length [2048]: 
  Enter private key passphrase []: 
  Rotate completed.
  
  '''
# ---
# name: test_rotate_with_password
  '''
  Create a new connection? [Y/n]: Y
  Enter connection name: keypairconnection
  Enter key length [2048]: 4096
  Enter private key passphrase []: 
  Set the `PRIVATE_KEY_PASSPHRASE` environment variable before using the connection.
  Rotate completed.
  
  '''
# ---
# name: test_setup_connection_already_exists
  '''
  Create a new connection? [Y/n]: 
  Enter connection name: default
  Enter key length [2048]: 
  Enter private key passphrase []: 
  +- Error ----------------------------------------------------------------------+
  | Connection with name default already exists.                                 |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_setup_create_output_directory_with_proper_privileges
  '''
  Create a new connection? [Y/n]: Y
  Enter connection name: keypairconnection
  Enter key length [2048]: 4096
  Enter private key passphrase []: 
  Setup completed.
  
  '''
# ---
# name: test_setup_error_if_any_public_key_is_set[KEY-KEY]
  '''
  Create a new connection? [Y/n]: Y
  Enter connection name: keypairconnection
  Enter key length [2048]: 4096
  Enter private key passphrase []: 
  +- Error ----------------------------------------------------------------------+
  | The public key is set already. Use the rotate command instead.               |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_setup_error_if_any_public_key_is_set[KEY-None]
  '''
  Create a new connection? [Y/n]: Y
  Enter connection name: keypairconnection
  Enter key length [2048]: 4096
  Enter private key passphrase []: 
  +- Error ----------------------------------------------------------------------+
  | The public key is set already. Use the rotate command instead.               |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_setup_error_if_any_public_key_is_set[None-KEY]
  '''
  Create a new connection? [Y/n]: Y
  Enter connection name: keypairconnection
  Enter key length [2048]: 4096
  Enter private key passphrase []: 
  +- Error ----------------------------------------------------------------------+
  | The public key is set already. Use the rotate command instead.               |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_setup_no_prompts
  '''
  Create a new connection? [Y/n]: 
  Enter connection name: keypairconnection
  Set the `PRIVATE_KEY_PASSPHRASE` environment variable before using the connection.
  Setup completed.
  
  '''
# ---
# name: test_setup_overwrite_connection
  '''
  Create a new connection? [Y/n]: n
  Enter key length [2048]: 
  Enter private key passphrase []: 
  Setup completed.
  
  '''
# ---
# name: test_setup_with_password
  '''
  Create a new connection? [Y/n]: Y
  Enter connection name: keypairconnection
  Enter key length [2048]: 4096
  Enter private key passphrase []: 
  Set the `PRIVATE_KEY_PASSPHRASE` environment variable before using the connection.
  Setup completed.
  
  '''
# ---
